File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/CBuilder/MasterDetailCardDemo/MasterDetailCardDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include "shellapi.hpp"
#pragma hdrstop

#include "MasterDetailCardDemoMain.h"
#include "FilmsDemoData.h"
#include "AboutDemoForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxCalendar"
#pragma link "cxCheckBox"
#pragma link "cxContainer"
#pragma link "cxDBEdit"
#pragma link "cxDBLookupComboBox"
#pragma link "cxDropDownEdit"
#pragma link "cxGridCardView"
#pragma link "cxGridDBCardView"
#pragma link "cxHyperLinkEdit"
#pragma link "cxLookupEdit"
#pragma link "cxMaskEdit"
#pragma link "cxMemo"
#pragma link "cxTextEdit"
#pragma link "cxDBLookupEdit"
#pragma link "cxNavigator"
#pragma link "cxDBNavigator"
#pragma link "cxLookAndFeels"
#pragma link "cxBlobEdit"
#pragma link "cxDataStorage"
#pragma link "cxGridCustomLayoutView"
#pragma link "cxLabel"
#pragma link "cxLookAndFeelPainters"
#pragma resource "*.dfm"
TMasterDetailCardDemoMainForm *MasterDetailCardDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TMasterDetailCardDemoMainForm::TMasterDetailCardDemoMainForm(TComponent* Owner)
  : TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TMasterDetailCardDemoMainForm::FormCreate(TObject *Sender)
{
	FilmsDemoDM->FilmsFiltered = false;
	FilmsDemoDM->cdsFilmsPersons->MasterSource = FilmsDemoDM->dsFilms;
	FilmsDemoDM->cdsFilmsPersons->IndexFieldNames = "FILMID";
	FilmsDemoDM->cdsFilmsPersons->MasterFields = "ID";
}
//---------------------------------------------------------------------------
void __fastcall TMasterDetailCardDemoMainForm::miGridClick(TObject *Sender)
{
  if (miGrid->Checked)
	SetGridMasterDetailStyle();
  else
	SetStandardMasterDetailStyle();
}
//---------------------------------------------------------------------------
void TMasterDetailCardDemoMainForm::SetStandardMasterDetailStyle(void)
{
	TcxGridDBDataController *AData;
	lvFilmsPersons->Visible = False;
	AData = cvFilmsPersons->DataController;
	AData->DataModeController->SmartRefresh = False;
	FilmsDemoDM->cdsFilmsPersons->MasterSource = FilmsDemoDM->dsFilms;
	pnlDetail->Visible = True;
	lblMaster->Visible = True;
	lblStyle->Caption = "Standard master-detail style";
}
//---------------------------------------------------------------------------

void TMasterDetailCardDemoMainForm::SetGridMasterDetailStyle(void)
{
	TcxGridDBDataController *AData;
	lvFilmsPersons->Visible = True;
	AData = cvFilmsPersons->DataController;
	AData->DataModeController->SmartRefresh = True;
	pnlDetail->Visible = False;
	FilmsDemoDM->cdsFilmsPersons->MasterSource = NULL;
	lblMaster->Visible = False;
	lblStyle->Caption = "ExpressQuantumGrid master-detail style";
}

File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/Delphi/MasterDetailCardDemo/MasterDetailCardDemoMain.pas
unit MasterDetailCardDemoMain;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, cxGridCustomTableView, cxGridTableView, cxGridDBTableView, DB,
  cxGridLevel, cxControls, cxGridCommon, cxGridCustomView, cxGrid,
  cxGridCardView, cxGridDBCardView, StdCtrls, ExtCtrls,
  cxDBData, ComCtrls, ToolWin, Menus, cxStyles, ImgList, DBCtrls,
  cxDropDownEdit, cxLookupEdit, cxDBLookupComboBox, cxMaskEdit, cxCalendar,
  cxDBEdit, cxHyperLinkEdit, cxContainer, cxEdit, cxTextEdit, cxMemo,
  ActnList, cxCheckBox, cxCustomData, cxGraphics, cxFilter, cxData,
  cxClasses, cxNavigator, cxDBLookupEdit, cxDBNavigator, cxDataStorage,
  cxBlobEdit, cxLookAndFeels, cxLookAndFeelPainters, cxGridCustomLayoutView,
  BaseForm, cxLabel;

type
  TMasterDetailCardDemoMainForm = class(TfmBaseForm)
    Grid: TcxGrid;
    lvFilms: TcxGridLevel;
    pnlDetail: TPanel;
    tvFilms: TcxGridDBTableView;
    colFilmsCaption: TcxGridDBColumn;
    colFilmsYear: TcxGridDBColumn;
    colFilmsTagline: TcxGridDBColumn;
    colFilmsPlotOutline: TcxGridDBColumn;
    colFilmsRuntime: TcxGridDBColumn;
    colFilmsPhoto: TcxGridDBColumn;
    lblMaster: TcxLabel;
    lblDetail: TcxLabel;
    lblStyle: TcxLabel;
    Bevel1: TBevel;
    cvFilmsPersons: TcxGridDBCardView;
    cvFilmsPersonsPersonLineID: TcxGridDBCardViewRow;
    cvFilmsPersonsBIOGRAPHY: TcxGridDBCardViewRow;
    cvFilmsPersonsBIRTHNAME: TcxGridDBCardViewRow;
    cvFilmsPersonsDATEOFBIRTH: TcxGridDBCardViewRow;
    cvFilmsPersonsFIRSTNAME: TcxGridDBCardViewRow;
    cvFilmsPersonsLOCATIONOFBIRTH: TcxGridDBCardViewRow;
    cvFilmsPersonsNICKNAME: TcxGridDBCardViewRow;
    cvFilmsPersonsSECONDNAME: TcxGridDBCardViewRow;
    cvFilmsPersonsHOMEPAGE: TcxGridDBCardViewRow;
    miOptions: TMenuItem;
    miGrid: TMenuItem;
    cvFilmsPersonsName: TcxGridDBCardViewRow;
    Panel1: TPanel;
    Label1: TcxLabel;
    cbOccupation: TcxDBLookupComboBox;
    Label2: TcxLabel;
    edFirstName: TcxDBTextEdit;
    Label8: TcxLabel;
    edSecondName: TcxDBTextEdit;
    Label9: TcxLabel;
    edNickName: TcxDBTextEdit;
    Label7: TcxLabel;
    edHomePage: TcxDBHyperLinkEdit;
    DBNavigator1: TcxDBNavigator;
    edBirthName: TcxDBTextEdit;
    Label6: TcxLabel;
    edLocationOfBirth: TcxDBTextEdit;
    Label4: TcxLabel;
    deDateOfBirth: TcxDBDateEdit;
    Label3: TcxLabel;
    meBiography: TcxDBMemo;
    Label5: TcxLabel;
    lvFilmsPersons: TcxGridLevel;
    chbMale: TcxDBCheckBox;
    cvFilmsPersonsGender: TcxGridDBCardViewRow;
    cxEditStyleController1: TcxEditStyleController;
    procedure miGridClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    procedure SetStandardMasterDetailStyle;
    procedure SetGridMasterDetailStyle;
  protected
    procedure LookAndFeelChanged(Sender: TcxLookAndFeel;
      AChangedValues: TcxLookAndFeelValues); override;
  end;

var
  MasterDetailCardDemoMainForm: TMasterDetailCardDemoMainForm;

implementation

{$R *.dfm}

uses
  cxGridDBDataDefinitions, FilmsDemoData, AboutDemoForm;

procedure TMasterDetailCardDemoMainForm.miGridClick(Sender: TObject);
begin
  tvFilms.Navigator.Visible := GetMenuItemChecked(Sender);
  if GetMenuItemChecked(Sender) then
    SetGridMasterDetailStyle
  else
    SetStandardMasterDetailStyle;
end;

procedure TMasterDetailCardDemoMainForm.LookAndFeelChanged(
  Sender: TcxLookAndFeel; AChangedValues: TcxLookAndFeelValues);
begin
  inherited LookAndFeelChanged(Sender, AChangedValues);
  UpdateTableViewStyleSheet(tvFilms);
  UpdateCardViewStyleSheet(cvFilmsPersons);
  if CanUseStyleSheet then
    cxEditStyleController1.Style.Color := $00F7EAD9
  else
    cxEditStyleController1.Style.AssignedValues := [];
end;

procedure TMasterDetailCardDemoMainForm.SetGridMasterDetailStyle;
var
  AData: TcxGridDBDataController;
begin
  pnlDetail.Visible := False;
  AData := cvFilmsPersons.DataController;
  AData.DataModeController.SmartRefresh := True;
  lvFilmsPersons.Visible := True;
  FilmsDemoDM.cdsFilmsPersons.MasterSource := nil;
  lblMaster.Visible := False;
  lblStyle.Caption := 'ExpressQuantumGrid master-detail style';
end;

procedure TMasterDetailCardDemoMainForm.SetStandardMasterDetailStyle;
var
  AData: TcxGridDBDataController;
begin
  lvFilmsPersons.Visible := False;
  AData := cvFilmsPersons.DataController;
  AData.DataModeController.SmartRefresh := False;
  FilmsDemoDM.cdsFilmsPersons.MasterSource := FilmsDemoDM.dsFilms;
  pnlDetail.Visible := True;
  lblMaster.Visible := True;
  lblStyle.Caption := 'Standard master-detail style';
end;

procedure TMasterDetailCardDemoMainForm.FormCreate(Sender: TObject);
begin
  inherited;
  FilmsDemoDM.FilmsFiltered := False;
  FilmsDemoDM.cdsFilmsPersons.MasterSource := FilmsDemoDM.dsFilms;
  FilmsDemoDM.cdsFilmsPersons.IndexFieldNames := 'FILMID';
  FilmsDemoDM.cdsFilmsPersons.MasterFields := 'ID';
end;

end.
